server/src/main/java/org/elasticsearch/features/NodeFeature.java

```diff
@@ -14,8 +14,13 @@
  * A feature published by a node.
  *
  * @param id        The feature id. Must be unique in the node.
+ * @param optional  {@code true} if nodes do not have to have this feature to join a cluster.
  */
-public record NodeFeature(String id) {
+public record NodeFeature(String id, boolean optional) {
```

Comment:
I don't like the idea of optional features - these are features for which `org.elasticsearch.cluster.ClusterFeatures#clusterHasFeature` might go from `true` back to `false` and the whole point is that we don't want features to regress like that. Do we really need this concept?

Reply:
Yes - if a plugin adds a feature, then it will then be impossible to remove the plugin without doing a full stop-start, else the restarted nodes without the plugin won't join the cluster without the plugin feature. Optional features let plugins be removed in a rolling restart.

Reply:
Hmm I think plugin removal is substantially more complicated than this. What if the plugin added some custom metadata to the cluster state which persists across restarts? How will we handle races between checking if the cluster supports a feature and then trying to use that feature if the feature can vanish underneath us? This will trip assertions today.

In any case, I don't think we need it for this PR (no plugins today currently add features) and `optional=true` certainly shouldn't be the default behaviour. Let's split this work out into a separate task.

Reply:
Makes sense, I've removed it from this PR